### ISSUE TITLE
The service "nelmio_api_doc.render_docs.html.asset" has a dependency on a non-existent service "twig.extension.assets"

### ISSUE THREAD

**Issue by BafS** [nelmio/NelmioApiDocBundle#1876]
https://github.com/nelmio/NelmioApiDocBundle/pull/1842 require a dependency on `twig.extension.assets` but it's not a dependency of nelmio. Could we make this dependency optional ?

Error message:
```
Uncaught Symfony\Component\DependencyInjection\Exception\ServiceNotFoundException: The service "nelmio_api_doc.render_docs.html.asset" has a dependency on a non-existent service "twig.extension.assets".
```

**Comment by Th3Mouk** [comment]
I experimented a similar error during upgrade from 4.5.0 to 4.6.0
```
The service "nelmio_api_doc.render_docs.html" has a dependency on a non-existent service "twig".
```

**Comment by Th3Mouk** [comment]
Duplicate of #1874

**Comment by GuilhemN** [comment]
Thank you for the report! Could you check whether https://github.com/nelmio/NelmioApiDocBundle/pull/1878 does indeed fix your issue?

**Comment by BafS** [comment]
It seems that the issue is still present.

```
composer req nelmio/api-doc-bundle:dev-WITHOUTTWIG
```

I rm -rf my cache to be sure, but then when the cache needs to be re-build:

`
PHP Fatal error:  Uncaught Symfony\Component\DependencyInjection\Exception\ServiceNotFoundException: The service "nelmio_api_doc.render_docs.html.asset" has a dependency on a non-existent service "twig.extension.assets". Did you mean one of these: "twig.extension.trans", "twig.extension.code", "twig.extension.routing", "twig.extension.yaml", "twig.extension.debug"? in /Users/bafs/Dev/ProtonTech/containers/webserver/repos/api/vendor/symfony/dependency-injection/Compiler/CheckExceptionOnInvalidReferenceBehaviorPass.php:86
`

**Comment by GuilhemN** [comment]
Do you have an old version of twig bundle installed?

This error seems to indicate that you are using an unsupported version of the twig bundle (<4.4).

**Comment by Gemorroj** [comment]
have same issue

**Comment by GuilhemN** [comment]
@Gemorroj is this still happening with 4.6.1? If not, do you have an old version of twig bundle installed?

**Comment by Gemorroj** [comment]
@GuilhemN yes. Still happening with 4.6.1.
```
php /bin/composer.phar update
Loading composer repositories with package information
Restricting packages listed in "symfony/symfony" to "5.3.*"
Updating dependencies
Lock file operations: 0 installs, 1 update, 0 removals
  - Upgrading nelmio/api-doc-bundle (v4.5.0 => v4.6.1)
Writing lock file
Installing dependencies from lock file (including require-dev)
Package operations: 0 installs, 1 update, 0 removals
  - Upgrading nelmio/api-doc-bundle (v4.5.0 => v4.6.1): Extracting archive
Generating optimized autoload files
composer/package-versions-deprecated: Generating version class...
composer/package-versions-deprecated: ...done generating version class
124 packages you are using are looking for funding.
Use the `composer fund` command to find out more!

What about running composer global require symfony/thanks && composer thanks now?
This will spread some love by sending a star to the GitHub repositories of your fellow package maintainers.

Run composer recipes at any time to see the status of your Symfony recipes.

Executing script cache:clear [KO]
 [KO]
Script cache:clear returned with error code 1
!!
!!  In CheckExceptionOnInvalidReferenceBehaviorPass.php line 86:
!!

!!    The service "nelmio_api_doc.render_docs.html.asset" has a dependency on a non-existent service "twig.extension.assets". Did you mean one o
!!    f these: "twig.extension.security", "twig.extension.dump", "twig.extension.trans", "twig.extension.code", "twig.extension.routing", "twig.
!!    extension.yaml", "twig.extension.weblink", "twig.extension.form"?

!!

!!
!!
Script @auto-scripts was called via post-update-cmd
```

```
php /bin/composer.phar show
beberlei/doctrineextensions             v1.3.0             A set of extensions to Doctrine 2 that add support for additional query functi...
composer/package-versions-deprecated    1.11.99.4          Composer plugin that provides efficient querying for installed package version...
composer/semver                         3.2.5              Semver library that offers utilities, version constraint parsing and validation.
composer/xdebug-handler                 2.0.2              Restarts a process without Xdebug.
doctrine/annotations                    1.13.2             Docblock Annotations Parser
doctrine/cache                          2.1.1              PHP Doctrine Cache library is a popular cache implementation that supports man...
doctrine/collections                    1.6.8              PHP Doctrine Collections library that adds additional functionality on top of ...
doctrine/common                         3.1.2              PHP Doctrine Common project is a library that provides additional functionalit...
doctrine/data-fixtures                  1.5.0              Data Fixtures for all Doctrine Object Managers
doctrine/dbal                           2.13.3             Powerful PHP database abstraction layer (DBAL) with many features for database...
doctrine/deprecations                   v0.5.3             A small layer on top of trigger_error(E_USER_DEPRECATED) or PSR-3 logging with...
doctrine/doctrine-bundle                2.4.x-dev 62a188c  Symfony DoctrineBundle
doctrine/doctrine-fixtures-bundle       3.4.0              Symfony DoctrineFixturesBundle
doctrine/doctrine-migrations-bundle     3.1.1              Symfony DoctrineMigrationsBundle
doctrine/event-manager                  1.1.1              The Doctrine Event Manager is a simple PHP event system that was built to be u...
doctrine/inflector                      2.0.3              PHP Doctrine Inflector is a small library that can perform string manipulation...
doctrine/instantiator                   1.4.0              A small, lightweight utility to instantiate objects in PHP without invoking th...
doctrine/lexer                          1.2.1              PHP Doctrine Lexer parser library that can be used in Top-Down, Recursive Desc...
doctrine/migrations                     3.2.1              PHP Doctrine Migrations project offer additional functionality on top of the d...
doctrine/orm                            2.9.x-dev 2575aa5  Object-Relational-Mapper for PHP
doctrine/persistence                    2.2.2              The Doctrine Persistence project is a set of shared interfaces and functionali...
doctrine/sql-formatter                  1.1.1              a PHP SQL highlighting library
dragonmantank/cron-expression           v3.1.0             CRON for PHP: Calculate the next or previous run date and determine if a CRON ...
dvdoug/boxpacker                        3.9.2              An implementation of the 3D (actually 4D) bin packing/knapsack problem (aka cr...
egulias/email-validator                 3.1.1              A library for validating emails against several RFCs
fresh/doctrine-enum-bundle              v7.3.4             Provides support of ENUM type for Doctrine2 in Symfony applications.
friendsofphp/php-cs-fixer               v3.1.0             A tool to automatically fix PHP code style
friendsofphp/proxy-manager-lts          v1.0.5             Adding support for a wider range of PHP versions to ocramius/proxy-manager
giggsey/libphonenumber-for-php          8.12.33            PHP Port of Google's libphonenumber
giggsey/locale                          2.0                Locale functions required by libphonenumber-for-php
laminas/laminas-code                    4.4.3              Extensions to the PHP Reflection API, static code scanning, and code generation
lcobucci/clock                          2.0.0              Yet another clock abstraction
lcobucci/jwt                            4.1.4              A simple library to work with JSON Web Token and JSON Web Signature
lexik/jwt-authentication-bundle         v2.13.0            This bundle provides JWT authentication for your Symfony REST API
martin-georgiev/postgresql-for-doctrine v1.5.2             Adds PostgreSQL enhancements to Doctrine. Provides support for JSON, JSONB and...
monolog/monolog                         2.3.4              Sends your logs to files, sockets, inboxes, databases and various web services
myclabs/deep-copy                       1.10.2             Create deep copies (clones) of your objects
namshi/jose                             7.2.3              JSON Object Signing and Encryption library for PHP.
nelmio/api-doc-bundle                   v4.6.1             Generates documentation for your REST API from annotations
nelmio/cors-bundle                      2.1.1              Adds CORS (Cross-Origin Resource Sharing) headers support in your Symfony appl...
nikic/php-parser                        v4.13.0            A PHP parser written in PHP
odolbeau/phone-number-bundle            v3.4.2             Integrates libphonenumber into your Symfony application
pagerfanta/pagerfanta                   v3.3.0             Pagination for PHP
phar-io/manifest                        2.0.3              Component for reading phar.io manifest information from a PHP Archive (PHAR)
phar-io/version                         3.1.0              Library for handling version information and constraints
php-cs-fixer/diff                       v2.0.2             sebastian/diff v3 backport support for PHP 5.6+
phpdocumentor/reflection-common         2.2.0              Common reflection classes used by phpdocumentor to reflect the code structure
phpdocumentor/reflection-docblock       5.2.2              With this component, a library can provide support for annotations via DocBloc...
phpdocumentor/type-resolver             1.5.0              A PSR-5 based resolver of Class names, Types and Structural Element Names
phpspec/prophecy                        1.14.0             Highly opinionated mocking framework for PHP 5.3+
phpstan/phpstan                         0.12.99            PHPStan - PHP Static Analysis Tool
phpstan/phpstan-doctrine                0.12.44            Doctrine extensions for PHPStan
phpstan/phpstan-phpunit                 0.12.22            PHPUnit extensions and rules for PHPStan
phpstan/phpstan-symfony                 0.12.44            Symfony Framework extensions and rules for PHPStan
phpunit/php-code-coverage               9.2.7              Library that provides collection, processing, and rendering functionality for ...
phpunit/php-file-iterator               3.0.5              FilterIterator implementation that filters files based on a list of suffixes.
phpunit/php-invoker                     3.1.1              Invoke callables with a timeout
phpunit/php-text-template               2.0.4              Simple template engine.
phpunit/php-timer                       5.0.3              Utility class for timing
phpunit/phpunit                         9.5.9              The PHP Unit Testing framework.
psr/cache                               2.0.0              Common interface for caching libraries
psr/container                           1.1.1              Common Container Interface (PHP FIG PSR-11)
psr/event-dispatcher                    1.0.0              Standard interfaces for event handling.
psr/link                                1.1.1              Common interfaces for HTTP links
psr/log                                 1.1.4              Common interface for logging libraries
sebastian/cli-parser                    1.0.1              Library for parsing CLI options
sebastian/code-unit                     1.0.8              Collection of value objects that represent the PHP code units
sebastian/code-unit-reverse-lookup      2.0.3              Looks up which function or method a line of code belongs to
sebastian/comparator                    4.0.6              Provides the functionality to compare PHP values for equality
sebastian/complexity                    2.0.2              Library for calculating the complexity of PHP code units
sebastian/diff                          4.0.4              Diff implementation
sebastian/environment                   5.1.3              Provides functionality to handle HHVM/PHP environments
sebastian/exporter                      4.0.3              Provides the functionality to export PHP variables for visualization
sebastian/global-state                  5.0.3              Snapshotting of global state
sebastian/lines-of-code                 1.0.3              Library for counting the lines of code in PHP source code
sebastian/object-enumerator             4.0.4              Traverses array structures and object graphs to enumerate all referenced objects
sebastian/object-reflector              2.0.4              Allows reflection of object attributes, including inherited and non-public ones
sebastian/recursion-context             4.0.4              Provides functionality to recursively process PHP variables
sebastian/resource-operations           3.0.3              Provides a list of PHP built-in functions that operate on resources
sebastian/type                          2.3.4              Collection of value objects that represent the types of the PHP type system
sebastian/version                       3.0.2              Library that helps with managing the version number of Git-hosted PHP projects
shapecode/cron-bundle                   dev-master eb70c70 This bundle provides scheduled execution of Symfony commands
symfony/amqp-messenger                  v5.3.7             Symfony AMQP extension Messenger Bridge
symfony/cache                           v5.3.7             Provides an extended PSR-6, PSR-16 (and tags) implementation
symfony/cache-contracts                 v2.4.0             Generic abstractions related to caching
symfony/config                          v5.3.4             Helps you find, load, combine, autofill and validate configuration values of a...
symfony/console                         v5.3.7             Eases the creation of beautiful and testable command line interfaces
symfony/debug-bundle                    v5.3.4             Provides a tight integration of the Symfony Debug component into the Symfony f...
symfony/dependency-injection            v5.3.7             Allows you to standardize and centralize the way objects are constructed in yo...
symfony/deprecation-contracts           v2.4.0             A generic function and convention to trigger deprecation notices
symfony/doctrine-bridge                 v5.3.7             Provides integration for Doctrine with various Symfony components
symfony/doctrine-messenger              v5.3.7             Symfony Doctrine Messenger Bridge
symfony/dotenv                          v5.3.7             Registers environment variables from a .env file
symfony/error-handler                   v5.3.7             Provides tools to manage errors and ease debugging PHP code
symfony/event-dispatcher                v5.3.7             Provides tools that allow your application components to communicate with each...
symfony/event-dispatcher-contracts      v2.4.0             Generic abstractions related to dispatching event
symfony/filesystem                      v5.3.4             Provides basic utilities for the filesystem
symfony/finder                          v5.3.7             Finds files and directories via an intuitive fluent interface
symfony/flex                            v1.15.4            Composer plugin for Symfony
symfony/form                            v5.3.7             Allows to easily create, process and reuse HTML forms
symfony/framework-bundle                v5.3.7             Provides a tight integration between Symfony components and the Symfony full-s...
symfony/http-client                     v5.3.7             Provides powerful methods to fetch HTTP resources synchronously or asynchronously
symfony/http-client-contracts           v2.4.0             Generic abstractions related to HTTP clients
symfony/http-foundation                 v5.3.7             Defines an object-oriented layer for the HTTP specification
symfony/http-kernel                     v5.3.7             Provides a structured process for converting a Request into a Response
symfony/intl                            v5.3.7             Provides a PHP replacement layer for the C intl extension that includes additi...
symfony/lock                            v5.3.4             Creates and manages locks, a mechanism to provide exclusive access to a shared...
symfony/mailer                          v5.3.4             Helps sending emails
symfony/maker-bundle                    v1.33.0            Symfony Maker helps you create empty commands, controllers, form classes, test...
symfony/mercure                         v0.5.3             Symfony Mercure Component
symfony/mercure-bundle                  v0.3.2             Symfony MercureBundle
symfony/messenger                       v5.3.7             Helps applications send and receive messages to/from other applications or via...
symfony/mime                            v5.3.7             Allows manipulating MIME messages
symfony/monolog-bridge                  v5.3.7             Provides integration for Monolog with various Symfony components
symfony/monolog-bundle                  v3.7.0             Symfony MonologBundle
symfony/options-resolver                v5.3.7             Provides an improved replacement for the array_replace PHP function
symfony/password-hasher                 v5.3.7             Provides password hashing utilities
symfony/phpunit-bridge                  v5.3.7             Provides utilities for PHPUnit, especially user deprecation notices management
symfony/polyfill-intl-grapheme          v1.23.1            Symfony polyfill for intl's grapheme_* functions
symfony/polyfill-intl-icu               v1.23.0            Symfony polyfill for intl's ICU-related data and classes
symfony/polyfill-intl-idn               v1.23.0            Symfony polyfill for intl's idn_to_ascii and idn_to_utf8 functions
symfony/polyfill-intl-normalizer        v1.23.0            Symfony polyfill for intl's Normalizer class and related functions
symfony/polyfill-php81                  v1.23.0            Symfony polyfill backporting some PHP 8.1+ features to lower PHP versions
symfony/polyfill-uuid                   v1.23.0            Symfony polyfill for uuid functions
symfony/process                         v5.3.7             Executes commands in sub-processes
symfony/property-access                 v5.3.7             Provides functions to read and write from/to an object or array using a simple...
symfony/property-info                   v5.3.7             Extracts information about PHP class' properties using metadata of popular sou...
symfony/proxy-manager-bridge            v5.3.4             Provides integration for ProxyManager with various Symfony components
symfony/rate-limiter                    v5.3.4             Provides a Token Bucket implementation to rate limit input and output in your ...
symfony/redis-messenger                 v5.3.7             Symfony Redis extension Messenger Bridge
symfony/routing                         v5.3.7             Maps an HTTP request to a set of configuration variables
symfony/runtime                         v5.3.4             Enables decoupling PHP applications from global state
symfony/security-bundle                 v5.3.7             Provides a tight integration of the Security component into the Symfony full-s...
symfony/security-core                   v5.3.7             Symfony Security Component - Core Library
symfony/security-csrf                   v5.3.4             Symfony Security Component - CSRF Library
symfony/security-guard                  v5.3.7             Symfony Security Component - Guard
symfony/security-http                   v5.3.7             Symfony Security Component - HTTP Integration
symfony/serializer                      v5.3.4             Handles serializing and deserializing data structures, including object graphs...
symfony/service-contracts               v2.4.0             Generic abstractions related to writing services
symfony/stopwatch                       v5.3.4             Provides a way to profile code
symfony/string                          v5.3.7             Provides an object-oriented API to strings and deals with bytes, UTF-8 code po...
symfony/translation-contracts           v2.4.0             Generic abstractions related to translation
symfony/twig-bridge                     v5.3.7             Provides integration for Twig with various Symfony components
symfony/twig-bundle                     v5.3.4             Provides a tight integration of Twig into the Symfony full-stack framework
symfony/uid                             v5.3.3             Provides an object-oriented API to generate and represent UIDs
symfony/validator                       v5.3.7             Provides tools to validate values
symfony/var-dumper                      v5.3.7             Provides mechanisms for walking through any arbitrary PHP variable
symfony/var-exporter                    v5.3.7             Allows exporting any serializable PHP data structure to plain PHP code
symfony/web-link                        v5.3.4             Manages links between resources
symfony/yaml                            v5.3.6             Loads and dumps YAML files
theseer/tokenizer                       1.2.1              A small library for converting tokenized PHP source code into XML and potentia...
twig/extra-bundle                       v3.3.3             A Symfony bundle for extra Twig extensions
twig/twig                               v3.3.3             Twig, the flexible, fast, and secure template language for PHP
webmozart/assert                        1.10.0             Assertions to validate method input/output with nice error messages.
zircote/swagger-php                     3.2.3              swagger-php - Generate interactive documentation for your RESTful API using ph...
```

**Comment by BafS** [comment]
I don't have the twig bundle installed, as it is not a dependency

```
$ composer show 'twig/*'
twig/twig v3.3.3 Twig, the flexible, fast, and secure template language for PHP
```

**Comment by dragoste** [comment]
@Gemorroj I think I just had the same issue as yours. The problem is that Twig Bundle removes the `twig.extension.assets` service if `symfony/asset` is not available.

```php
# vendor/symfony/twig-bundle/DependencyInjection/Compiler/ExtensionPass.php
if (!$container::willBeAvailable('symfony/asset', Packages::class, ['symfony/twig-bundle'])) {
     $container->removeDefinition('twig.extension.assets');
}
```

So the quick fix is to install `symfony/asset` first.

Not having Twig Bundle at all is probably a slightly different case, so I guess it's not a solution for BafS.

**Comment by BafS** [comment]
I would be grateful if we don't have to install the bundle because for API only twig is not used, it could be a "suggested" requirement.

**Comment by enumag** [comment]
After upgrading from v4.5 to v4.6 I'm getting this error. A bit different but probably related?

```
Symfony\Component\DependencyInjection\Exception\ServiceNotFoundException: The service "twig.extension.assets" has a dependency on a non-existent service "assets.packages".
```

**Comment by mkldevops** [comment]
> @Gemorroj I think I just had the same issue as yours. The problem is that Twig Bundle removes the `twig.extension.assets` service if `symfony/asset` is not available.
> 
> ```
> # vendor/symfony/twig-bundle/DependencyInjection/Compiler/ExtensionPass.php
> if (!$container::willBeAvailable('symfony/asset', Packages::class, ['symfony/twig-bundle'])) {
>      $container->removeDefinition('twig.extension.assets');
> }
> ```
> 
> So the quick fix is to install `symfony/asset` first.
> 
> Not having Twig Bundle at all is probably a slightly different case, so I guess it's not a solution for BafS.

It's works thanks

**Comment by geoffroyp** [comment]
Bump. I have a fresh install a ran into this problem today, so it's still there...

**Comment by GuilhemN** [comment]
Sorry I did not have time to fix this myself, thanks a lot to @Gemorroj who fixed this in #1919!
